cpu/stm32/periph/flashpage.c

```diff
@@ -253,3 +319,54 @@ void flashpage_write(void *target_addr, const void *data, size_t len)
     }
 #endif
 }
+
+#if !(defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || \
+      defined(CPU_FAM_STM32F7))
+void flashpage_write_page(unsigned page, const void *data)
+{
+    flashpage_erase(page);
+
+    /* WRITE sequence */
+    if (data != NULL) {
+        flashpage_write(flashpage_addr(page), data, FLASHPAGE_SIZE);
+    }
+}
+#endif
+
+#if defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || \
+    defined(CPU_FAM_STM32F7)
+size_t flashpage_size(unsigned page)
+{
+    if (page < 4) {
+        return 16 * 1024;
+    }
+    else if (page == 4) {
+        return 64 * 1024;
+    }
+    else {
+        return 128 * 1024;
+    }
+}
+
+void *flashpage_addr(unsigned page)
+{
+    if (page == 0) {
+        return (void*)CPU_FLASH_BASE;
+    }
+    return (void*)((intptr_t)flashpage_addr(page - 1) + flashpage_size(page - 1));
+}
+
+unsigned flashpage_page(void *addr)
+{
+    unsigned page = (((intptr_t)addr - CPU_FLASH_BASE) / (16 * 1024));
+
+    if (page > 7) {
+        page = (page / 8) + 4;
+    }
+    else if (page > 3) {
+        page = 4;
+    }
```

Comment:
it's not obvious to me how that works

Reply:
If we want to provide a generic function we could do

```C
unsigned flashpage_page(void *addr)
{
    unsigned page = 0;

    for (uintptr_t pos = 0; addr > pos; ++page) {
        pos += flashpage_size(page);
    }

    return page - 1;
}
```

Reply:
Heavily commented the function, let me know if it is more clear now